### ISSUE TITLE
Repeating digital objects when script encounters missing images

### ISSUE THREAD

**Issue by clingerman** [usnationalarchives/partner-data-transform#1]
@DominicBM When the script encounters images that are listed in the original partner metadata but not in the objects CSV, it repeats the previous valid file name from the CSV as many times as it takes to get to the next match between the two. Example below repeats 2777.jpg:

```
<digitalObject><objectType><termName>Image (JPG)</termName></objectType><labelFlag>2776.jpg</labelFlag>
<objectDesignator>Fold3 File #8795155</objectDesignator>
<objectDescription>Image provided by Fold3.</objectDescription>
<accessFilename>https://opaexport-conv.s3.amazonaws.com/TB0104/M1676-NaturalizationIndex-NYSouthernPetitions/M1676_006/images/2776.jpg</accessFilename><accessFileSize>1522921</accessFileSize>
<thumbnailFilename>http://media.nara.gov/dc-metro/jpg_t.jpg</thumbnailFilename><thumbnailFileSize>1234</thumbnailFileSize></digitalObject>

<digitalObject><objectType><termName>Image (JPG)</termName></objectType><labelFlag>2777.jpg</labelFlag>
<objectDesignator>Fold3 File #8795159</objectDesignator>
<objectDescription>Image provided by Fold3.</objectDescription>
<accessFilename>https://opaexport-conv.s3.amazonaws.com/TB0104/M1676-NaturalizationIndex-NYSouthernPetitions/M1676_006/images/2777.jpg</accessFilename><accessFileSize>1493678</accessFileSize>
<thumbnailFilename>http://media.nara.gov/dc-metro/jpg_t.jpg</thumbnailFilename><thumbnailFileSize>1234</thumbnailFileSize></digitalObject>

<digitalObject><objectType><termName>Image (JPG)</termName></objectType><labelFlag>2777.jpg</labelFlag>
<objectDesignator>Fold3 File #8729037</objectDesignator>
<objectDescription>Image provided by Fold3.</objectDescription>
<accessFilename>https://opaexport-conv.s3.amazonaws.com/TB0104/M1676-NaturalizationIndex-NYSouthernPetitions/M1676_006/images/2777.jpg</accessFilename><accessFileSize></accessFileSize>
<thumbnailFilename>http://media.nara.gov/dc-metro/jpg_t.jpg</thumbnailFilename><thumbnailFileSize>1234</thumbnailFileSize></digitalObject>

<digitalObject><objectType><termName>Image (JPG)</termName></objectType><labelFlag>2777.jpg</labelFlag>
<objectDesignator>Fold3 File #8834153</objectDesignator>
<objectDescription>Image provided by Fold3.</objectDescription>
<accessFilename>https://opaexport-conv.s3.amazonaws.com/TB0104/M1676-NaturalizationIndex-NYSouthernPetitions/M1676_006/images/2777.jpg</accessFilename><accessFileSize></accessFileSize>
<thumbnailFilename>http://media.nara.gov/dc-metro/jpg_t.jpg</thumbnailFilename><thumbnailFileSize>1234</thumbnailFileSize></digitalObject>

<digitalObject><objectType><termName>Image (JPG)</termName></objectType><labelFlag>2777.jpg</labelFlag>
<objectDesignator>Fold3 File #8834173</objectDesignator>
<objectDescription>Image provided by Fold3.</objectDescription>
<accessFilename>https://opaexport-conv.s3.amazonaws.com/TB0104/M1676-NaturalizationIndex-NYSouthernPetitions/M1676_006/images/2777.jpg</accessFilename><accessFileSize></accessFileSize>
<thumbnailFilename>http://media.nara.gov/dc-metro/jpg_t.jpg</thumbnailFilename><thumbnailFileSize>1234</thumbnailFileSize></digitalObject>
```

This is happening because the objects CSV has a gap between 2777.jpg and 3171.jpg (see below):

```
TB0104/M1676-NaturalizationIndex-NYSouthernPetitions/M1676_006/images/2776.jpg	"1522921"	"TB0104"	"M1676-NaturalizationIndex-NYSouthernPetitions"	"M1676_0006"	"images"	"2776.jpg"

TB0104/M1676-NaturalizationIndex-NYSouthernPetitions/M1676_006/images/2777.jpg	"1493678"	"TB0104"	"M1676-NaturalizationIndex-NYSouthernPetitions"	"M1676_0006"	"images"	"2777.jpg"

TB0104/M1676-NaturalizationIndex-NYSouthernPetitions/M1676_006/images/3171.jpg	"1421791"	"TB0104"	"M1676-NaturalizationIndex-NYSouthernPetitions"	"M1676_0006"	"images"	"3171.jpg"
```

Is there anyway we could modify the script so that when it encounters images in the original partner data that aren't in the objects CSV, that it creates a description without any digital objects but still with the metadata?

Relevant files for this reel (M1676_0006) are attached.
[M1676_0006.zip](https://github.com/usnationalarchives/partner-data-transform/files/968543/M1676_0006.zip)

**Comment by DominicBM** [comment]
I can see the problem is that the script was written assuming every object has a match. It iterates through each file name, overwriting the previous file's data fields with the next one. But this process means that if none is found, it just leaves the previous file's data intact for the next iteration.

I think it should be possible to add some code that will follow different steps if there is no match.

**Comment by DominicBM** [comment]
@clingerman Okay, I think I have figured out what is going on. There were actually two different issues happening. First, the script tries to create data for a digital object from each "page" item in the metadata. It didn't have any checking built in, because it assumed it would always find object data and overwrite the variables from the previous object with the current. When it wasn't finding a match for the file name, it would still write an object as if it had, using the previous data. I have introduced a test where it wipes the values after each iteration, and then will only try to write a new object if `file_path` has a value (so we know it belongs to the current one). That should solve the problem of writing objects where there is no data for it.

There was actually another problem, which was that the script wrote the 2777.jpg object repeatedly in one file unit, rather than creating a new file unit for each one. This means that it was for some reasoning determining that that the titles were not unique, even though you can look at the original XML and see that the names do change between 2777.jpg and 3171.jpg. I realized that that's because some of these names are repeats. Newman Dubrovsky, for example, is 2778.jpg, but he's already had a file unit created for 2201.jpg.

This presents a problem, since in the original data we were working with, the files related to the same file unit were all in sequence in the data, or at least that was the assumption. This problem may be causing other files to become attached to the wrong file unit as well, since the script was designed to simply attach an object to a file unit if the title is not unique, because it assumed the title is not unique due to matching the previous object's file unit. Now the question is, **do we want to put all the Newman Dubrovskys in the data set into one file unit, regardless of their order in the original data?** If so, we'd need to introduce a step that sorts the original XML before transforming it, I think. (This will get complicated if the same title can appear across rolls...) Or do we want to create a new file unit if Dubrovsky appears multiple times non-contiguously in the data? To make that happen, we'd have to just check for uniqueness against the last title, rather than all titles. That would be easiest, but would end up possibly creating multiple file units with the same name, and I'm not sure if that's what you want.

Let me know if that makes sense. It's hard to explain in words; we can discuss further if you want.

**Comment by clingerman** [comment]
I don't think we should put all the Dubrovsky's in the same file unit, but instead respect "original order." I think the repeats of names in this particular roll were a mistake on the partner's part - i.e. duplication that they "caught" by removing the objects they delivered us but not removing the duplicate metadata for the removed objects - thus why Dubrovsky has 2201.jpg and 2778.jpg.

What would be **really** helpful would be if there was some way for the script to generate a list images that are missing for us to refer back to.

**Comment by eboudreau** [comment]
Attaching some sample files - an objects CSV, its matching metadata XML, and the python script.

[BlankImageSample.zip](https://github.com/usnationalarchives/partner-data-transform/files/1974993/BlankImageSample.zip)

**Comment by ghstern** [comment]
"Scraping missing JPGs from Fold3"

Also, can we get there should be a script that:
1) takes the fold3 objectDesignator from the metadata xml for the missing JPG
2) goes to the Fold3 site
3) downloads a copy to a directory (locally?)
4) Writes a new URL for that object in the description
        perhaps in a folder called "scraped" so we don't have to worry about putting them into the directory that is missing this particular file.

Please let me know if you have any questions.
Thanks!
